File: snippets/python/zfunc.py
def z_function(s: str) -> list[int]:
    """
    z[i] = length of longest substring starting at i which is also a prefix
    Time: O(n)
    """
    n = len(s)
    z = [0] * n
    z[0] = n
    l = r = 0
    
    for i in range(1, n):
        if i < r:
            z[i] = min(r - i, z[i - l])
        while i + z[i] < n and s[z[i]] == s[i + z[i]]:
            z[i] += 1
        if i + z[i] > r:
            l, r = i, i + z[i]
    return z

def z_search(text: str, pattern: str) -> list[int]:
    """Find all occurrences of pattern in text using Z-algorithm"""
    if not pattern:
        return []
    concat = pattern + "$" + text
    z = z_function(concat)
    m = len(pattern)
    return [i - m - 1 for i in range(m + 1, len(concat)) if z[i] >= m]

File: snippets/rust/strings/zfunc.rs
fn z_function(s: &[u8]) -> Vec<usize> {
    let n = s.len();
    let mut z = vec![0usize; n];
    if n == 0 { return z; }
    z[0] = n;
    let mut l = 0usize;
    let mut r = 0usize;
    for i in 1..n {
        if i <= r { z[i] = (r - i + 1).min(z[i - l]); }
        while i + z[i] < n && s[z[i]] == s[i + z[i]] { z[i] += 1; }
        if z[i] > 0 && i + z[i] - 1 > r { l = i; r = i + z[i] - 1; }
    }
    z
}
